core/src/banking_stage/latest_unprocessed_votes.rs

```diff
@@ -274,49 +274,41 @@ impl LatestUnprocessedVotes {
         bank: Arc<Bank>,
         forward_packet_batches_by_accounts: &mut ForwardPacketBatchesByAccounts,
     ) -> usize {
-        let mut continue_forwarding = true;
-        let pubkeys_by_stake = weighted_random_order_by_stake(
-            &bank,
-            self.latest_votes_per_pubkey.read().unwrap().keys(),
-        )
-        .collect_vec();
-        pubkeys_by_stake
-            .into_iter()
-            .filter(|&pubkey| {
-                if !continue_forwarding {
-                    return false;
-                }
-                if let Some(lock) = self.get_entry(pubkey) {
-                    let mut vote = lock.write().unwrap();
-                    if !vote.is_vote_taken() && !vote.is_forwarded() {
-                        let deserialized_vote_packet = vote.vote.as_ref().unwrap().clone();
-                        if let Some(sanitized_vote_transaction) = deserialized_vote_packet
-                            .build_sanitized_transaction(
-                                bank.vote_only_bank(),
-                                bank.as_ref(),
-                                bank.get_reserved_account_keys(),
-                            )
-                        {
-                            if forward_packet_batches_by_accounts.try_add_packet(
-                                &sanitized_vote_transaction,
-                                deserialized_vote_packet,
-                                &bank.feature_set,
-                            ) {
-                                vote.forwarded = true;
-                            } else {
-                                // To match behavior of regular transactions we stop
-                                // forwarding votes as soon as one fails
-                                continue_forwarding = false;
-                            }
-                            return true;
-                        } else {
-                            return false;
-                        }
-                    }
-                }
-                false
-            })
-            .count()
+        let binding = self.latest_votes_per_pubkey.read().unwrap();
+        let pubkeys_by_stake = weighted_random_order_by_stake(&bank, binding.keys());
```

Comment:
I think this is locking the `latest_votes_per_pubkey` for longer than necessary.
If the following compiles, it is a slightly better way to do it, as it releases the lock earlier:

```suggestion
        let pubkeys_by_stake = {
            let binding = self.latest_votes_per_pubkey.read().unwrap();
            weighted_random_order_by_stake(&bank, binding.keys())
        };
```

Reply:
This is because `weighted_random_order_by_stake` return type incorrectly captures the lifetime of the second argument.
I've suggested how you can fix it [above](https://github.com/anza-xyz/agave/pull/2071#discussion_r1671908547).

Here is a working code with the lock lifetime reduced and some other fixes I've suggested: https://github.com/anza-xyz/agave/pull/2091